ribbon-core/src/main/java/com/netflix/client/config/AbstractReloadableClientConfig.java

```diff
@@ -0,0 +1,389 @@
+    }
+
+    private <T> ReloadableProperty<T> getClientDynamicProperty(IClientConfigKey<T> key, boolean isDynamic) {
+        LOG.debug("Get dynamic property key={} ns={} client={}", key.key(), getNameSpace(), clientName);
+
+        return createProperty(
+                () -> resolveFinalProperty(key),
+                key::defaultValue,
+                isDynamic);
+    }
+
+    /**
+     * Resolve a properties final value in the following order or precedence
+     * - client scope
+     * - default scope
+     * - internally set default
+     * - IClientConfigKey defaultValue
+     * @param key
+     * @param <T>
+     * @return
+     */
+    private <T> Optional<T> resolveFinalProperty(IClientConfigKey<T> key) {
+        Optional<T> value;
+        if (!StringUtils.isEmpty(clientName)) {
+            value = getFromStorage(clientName + "." + getNameSpace() + "." + key.key(), key.type());
+            if (value.isPresent()) {
+                return value;
+            }
+        }
+
+        value = getFromStorage(getNameSpace() + "." + key.key(), key.type());
+        if (value.isPresent()) {
+            return value;
+        }
+
+        value = resolveDefaultProperty(key);
+        if (value.isPresent()) {
+            return value;
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Returns the internal property to the desiredn type
+     */
+    protected <T> Optional<T> resolveDefaultProperty(IClientConfigKey<T> key) {
+        return Optional.ofNullable(defaultProperties.get(key.key()))
+                .map(value -> {
+                    final Class type = key.type();
+                    // Unfortunately there's some legacy code setting string values for typed keys.  Here are do our best to parse
+                    // and store the typed value
+                    if (!value.getClass().equals(type)) {
+                        if (value.getClass().equals(String.class)) {
+                            final String strValue = (String) value;
+                            if (Integer.class.equals(type)) {
+                                return (T)Integer.valueOf(strValue);
+                            } else if (Boolean.class.equals(type)) {
+                                return (T)Boolean.valueOf(strValue);
+                            } else if (Float.class.equals(type)) {
+                                return(T) Float.valueOf(strValue);
+                            } else if (Long.class.equals(type)) {
+                                return (T)Long.valueOf(strValue);
+                            } else if (Double.class.equals(type)) {
+                                return (T)Double.valueOf(strValue);
+                            } else if (TimeUnit.class.equals(type)) {
+                                return (T)TimeUnit.valueOf(strValue);
+                            } else {
+                                LOG.warn("Unsupported value type `{}` for key `{}` with value `{}`",
```

Comment:
should you do something similar in a catch Exception block in case one of these parse methods fails?

Reply:
Good point.  I'm just going to make this an error.